Sources/PrivacyStats/internal/PrivacyStatsUtils.swift

```diff
@@ -0,0 +1,121 @@
+
+    /**
+     * Returns a dictionary representation of blocked trackers counts grouped by company name for the current day.
+     */
+    static func loadCurrentDayStats(in context: NSManagedObjectContext) throws -> [String: Int64] {
+        let startDate = Date.currentPrivacyStatsPackTimestamp
+        return try loadBlockedTrackersStats(since: startDate, in: context)
+    }
+
+    /**
+     * Returns a dictionary representation of blocked trackers counts grouped by company name for past 7 days.
+     */
+    static func load7DayStats(in context: NSManagedObjectContext) throws -> [String: Int64] {
+        let startDate = Date().privacyStatsOldestPackTimestamp
+        return try loadBlockedTrackersStats(since: startDate, in: context)
+    }
+
+    private static func loadBlockedTrackersStats(since startDate: Date, in context: NSManagedObjectContext) throws -> [String: Int64] {
+        let request = NSFetchRequest<NSFetchRequestResult>(entityName: DailyBlockedTrackersEntity.Const.entityName)
+        request.predicate = NSPredicate(format: "%K >= %@", #keyPath(DailyBlockedTrackersEntity.timestamp), startDate as NSDate)
+
+        let companyNameKey = #keyPath(DailyBlockedTrackersEntity.companyName)
+
+        // Expression description for the sum of count
+        let countExpression = NSExpression(forKeyPath: #keyPath(DailyBlockedTrackersEntity.count))
+        let sumExpression = NSExpression(forFunction: "sum:", arguments: [countExpression])
+
+        let sumExpressionDescription = NSExpressionDescription()
+        sumExpressionDescription.name = "totalCount"
+        sumExpressionDescription.expression = sumExpression
+        sumExpressionDescription.expressionResultType = .integer64AttributeType
+
+        request.propertiesToGroupBy = [companyNameKey]
+        request.propertiesToFetch = [companyNameKey, sumExpressionDescription]
+        request.resultType = .dictionaryResultType
+
+        let results = (try context.fetch(request) as? [[String: Any]]) ?? []
+
+        let groupedResults = results.reduce(into: [String: Int64]()) { partialResult, result in
+            if let companyName = result[companyNameKey] as? String, let totalCount = result["totalCount"] as? Int64, totalCount > 0 {
+                partialResult[companyName] = totalCount
+            }
+        }
+
+        return groupedResults
+    }
+
+    /**
+     * Deletes stats older than 7 days for all companies.
+     */
+    static func deleteOutdatedPacks(in context: NSManagedObjectContext) throws {
+        let oldestValidTimestamp = Date().privacyStatsOldestPackTimestamp
+
+        let fetchRequest = NSFetchRequest<NSFetchRequestResult>(entityName: DailyBlockedTrackersEntity.Const.entityName)
+        fetchRequest.predicate = NSPredicate(format: "%K < %@", #keyPath(DailyBlockedTrackersEntity.timestamp), oldestValidTimestamp as NSDate)
+        let deleteRequest = NSBatchDeleteRequest(fetchRequest: fetchRequest)
+
+        try context.execute(deleteRequest)
```

Comment:
As we are using batch delete request, I'd reset the context after this to invalidate any object that could be still cached in the object graph by the context.